libbeat/feature/registry.go

```diff
@@ -0,0 +1,141 @@
+package feature
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+type mapper map[string]map[string]Featurable
+
+// Registry implements a global registry for any kind of feature in beats.
+// feature are grouped by namespace, a namespace is a kind of plugin like outputs, inputs, or queue.
+// The feature name must be unique.
+type registry struct {
+	sync.RWMutex
+	namespaces mapper
+	log        *logp.Logger
+}
+
+// NewRegistry returns a new registry.
+func newRegistry() *registry {
+	return &registry{
+		namespaces: make(mapper),
+		log:        logp.NewLogger("registry"),
+	}
+}
+
+// Register registers a new feature into a specific namespace, namespace are lazy created.
+// Feature name must be unique.
+func (r *registry) Register(feature Featurable) error {
+	r.Lock()
+	defer r.Unlock()
+
+	// Lazy create namespaces
+	_, found := r.namespaces[feature.Namespace()]
+	if !found {
+		r.namespaces[feature.Namespace()] = make(map[string]Featurable)
+	}
+
+	f, found := r.namespaces[feature.Namespace()][feature.Name()]
+	if found {
+		if feature.Equal(f) {
+			// Allow both old style and new style of plugin to work together.
+			r.log.Debugw(
+				"ignoring, feature '%s' is already registered in the namespace '%s'",
+				feature.Name(),
+				feature.Namespace(),
+			)
+			return nil
+		}
+
+		return fmt.Errorf(
+			"could not register new feature '%s' in namespace '%s', feature name must be unique",
+			feature.Name(),
+			feature.Namespace(),
+		)
+	}
+
+	r.log.Debugw(
+		"registering new feature",
+		"namespace",
+		feature.Namespace(),
+		"name",
+		feature.Name(),
+	)
+
+	r.namespaces[feature.Namespace()][feature.Name()] = feature
```

Comment:
Instead of calling a function that might potentially generate a string on every call (it's all interfaces), let's introduce some local vars: `ns, name := feature.Namespace(), feature.Name()`.